gophermart

msmkdenis opened the pull request.

msmkdenis committed:
initial commit

msmkdenis committed:
test pr

msmkdenis committed:
fix tests

msmkdenis committed:
go mod init

msmkdenis committed:
feat: user register login

msmkdenis committed:
fix: linters, self autotests

msmkdenis committed:
feat: orders upload getall

msmkdenis committed:
fix: get api/user/orders url

msmkdenis committed:
feat: order accrual system

msmkdenis committed:
feat: balance withdrawal

msmkdenis committed:
fix: withdraw endpoint

msmkdenis committed:
test: balancehandler tests

msmkdenis committed:
feat: add docker compose and mock accrual system for local tests, to start - docker compose up from root

rktkov commented on internal/utils/jwt_manager.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+type JWTManager struct {
+	logger    *zap.Logger
```
Я бы для логгирования сделал бы внутренний пакет, с необходимым интерфейсом, а внутри пакета уже использовал бы библиотеку. Так будет больше гибкости в ситуации, если захочется расширить инструменты логгирования.

msmkdenis replied:
Понимаю причину - позволит нам в одном месте поменять логгер на любой другой, не меня код в дальнейшем в теле проекта, т.к. он будет завязан на внутренние интерфейсы. Если возможно оставил бы пока такую реализацию.

rktkov commented on internal/user/userrepository/user_repository.go:
```diff
@@ -0,0 +1,62 @@
+var selectUserByLogin string
+
+type PostgresUserRepository struct {
+	PostgresPool *db.PostgresPool
```
А точно нам нужно поле делать открытым/доступным вне структуры.

msmkdenis replied:
Скорректировал здесь и в других пакетах для аналогичных случаев

rktkov commented on internal/user/userhandler/user_handler.go:
```diff
@@ -0,0 +1,145 @@
+package userhandler
```
Я бы пакеты переименовал, убрав слово user. Так и дерево проекта будет чище и лишнее повторяющееся слово уйдет. И тоже самое для пакетов других сущностей сделал бы.

msmkdenis replied:
Скорректировал

rktkov commented on internal/order/orderservice/order_accrual_service.go:
```diff
@@ -0,0 +1,93 @@
+	rl.Take()
+	updatedOrder, err := oc.queryAccrual.QueryUpdateOrder(o.Number)
+
+	if err == nil {
```
Более привычно, когда первым рассматривают случай когда ошибка не nil. А потом уже идет код в основном теле функции в ситуации, когда ошибки нет. Стоит перенести код с 88 строчки сюда.

msmkdenis replied:
Скорректировал

rktkov commented on internal/order/orderservice/order_accrual_service.go:
```diff
@@ -0,0 +1,93 @@
+			oc.logger.Error("error while updating order", zap.Error(errUpdateOrder))
+		}
+
+		errUpdateBalance := oc.balanceRepository.UpdateBalance(context.Background(), o.UserLogin, o.Accrual)
```
Я точно хочу делать обновление баланса, если у меня выше произошла ошибка обновления заказа? Очень похоже, что могу консистентность данных порушить.
Тут вообще подумать, как два действия объединить в одно контролируемое. Т.е. если хотя бы в одном запросе, что пойдет не так, откатить уже выполненные действия.

msmkdenis replied:
Вынес логику взаимодействия с accrual системой в отдельный пакет. Например - вдруг у нас будут появляться новые способы взаимодействия с accrual системой.
Согласен, мне тоже не нравилось если честно.
Обновление order  и balance сделал в рамках одной транзакции.

msmkdenis replied:
Решил скорректировать подход. В описанном выше бизнес логика "протекала" в слой репозитория. При этом чтобы не запускать транзакции в сервисе приходилось делать отдельный репозиторий, который вызывал методы более подходящие другим доменам (и даже дублирующиеся в других репозиториях). 
Добавил transaction manager для управления транзакциями из слоя сервиса. Убрал "лишний репозиторий". В transaction manager запускаю методы из разных репозиториев (в данном случае из order и balance репозиториев). Т.е. "бизнесовой" транзакцией (обновить одновременно начисление на заказ и общий баланс) теперь управляет слой сервиса.

rktkov commented on internal/order/orderservice/order_service.go:
```diff
@@ -0,0 +1,71 @@
+	"context"
+	"fmt"
+
+	"github.com/msmkdenis/yap-gophermart/internal/apperrors"
```
Это осознанное обособление инициализации пакета ошибок отдельно от других внутренних пакетов?

msmkdenis replied:
Скорректировал, так не должно было быть

rktkov commented on internal/middleware/request-logger.go:
```diff
@@ -0,0 +1,80 @@
+	RequestLogger struct {
+		ReqLogger *zap.Logger
+	}
+	// берём структуру для хранения сведений об ответе
```
В других местах кода я видел комментарии на английском языке. Надо выбрать один язык и использовать его во всем коде.

msmkdenis replied:
Убрал англоязычные комментарии

rktkov commented on internal/app/gophermart/app.go:
```diff
@@ -0,0 +1,121 @@
+
+	serverCtx, serverStopCtx := context.WithCancel(context.Background())
+
+	quit := make(chan os.Signal, 1)
```
Я бы обработку сигналов о завершении процесса из ОС вынес бы в main.

msmkdenis replied:
канал с ос сигналами вынес в main, передаю его в run и там обрабатываю

rktkov commented on internal/app/gophermart/app.go:
```diff
@@ -0,0 +1,121 @@
+	go func() {
+		<-quit
+
+		// Shutdown signal with grace period of 30 seconds
```
Вот комментарий на английском.

msmkdenis replied:
Убрал

msmkdenis committed:
fix: review_1

msmkdenis committed:
fix: vet

msmkdenis committed:
fix: accrual

msmkdenis committed:
fix: linters

msmkdenis committed:
fix: add avito transaction manager

add tx manager to manage different repos functions
from service layer within one transaction with certain options

msmkdenis committed:
fix: negative balance violation error, withdraw  validation

add specific check for negative balance constraint violation
add custom dto validation for not negative withdraw

msmkdenis committed:
fix: static test

msmkdenis commented:
Поправил основные проблемные места + нашел у себя некорректную обработку ошибки constraint-а из ДБ: ошибка при некорректном withdrawal показывалась как ошибка отрицательного баланса. 
Сделал отдельную проверку на именованный constraint, сделал валидацию на положительную сумму списания в хендлере (в базу больше с отрицательным списанием в запросе не пускает, ошибка на уровне хендлера).

msmkdenis committed:
fix: os signals moved to main
